crates/goose/src/agents/todo_extension.rs

```diff
@@ -0,0 +1,220 @@
+                version: "1.0.0".to_string(),
+            },
+            instructions: Some(
+                "Manage TODO lists - read and write task planning content.".to_string(),
+            ),
+        };
+
+        Ok(Self { info, context })
+    }
+
+    async fn handle_read_todo(&self) -> Result<Vec<Content>, String> {
+        match SessionManager::get_session(&self.context.session_id, false).await {
+            Ok(metadata) => {
+                let content =
+                    extension_data::TodoState::from_extension_data(&metadata.extension_data)
+                        .map(|state| state.content)
+                        .unwrap_or_default();
+                Ok(vec![Content::text(content)])
+            }
+            Err(_) => Ok(vec![Content::text(String::new())]),
+        }
+    }
+
+    async fn handle_write_todo(&self, arguments: Value) -> Result<Vec<Content>, String> {
+        let content = arguments
+            .get("content")
+            .and_then(|v| v.as_str())
+            .ok_or("Missing required parameter: content")?
+            .to_string();
+
+        let char_count = content.chars().count();
+        let max_chars = std::env::var("GOOSE_TODO_MAX_CHARS")
+            .ok()
+            .and_then(|s| s.parse().ok())
+            .unwrap_or(50_000);
+
+        if max_chars > 0 && char_count > max_chars {
+            return Err(format!(
+                "Todo list too large: {} chars (max: {})",
+                char_count, max_chars
+            ));
+        }
+
+        match SessionManager::get_session(&self.context.session_id, false).await {
+            Ok(mut session) => {
+                let todo_state = extension_data::TodoState::new(content);
+                if todo_state
+                    .to_extension_data(&mut session.extension_data)
+                    .is_ok()
+                {
+                    match SessionManager::update_session(&self.context.session_id)
+                        .extension_data(session.extension_data)
+                        .apply()
+                        .await
+                    {
+                        Ok(_) => Ok(vec![Content::text(format!(
+                            "Updated ({} chars)",
+                            char_count
+                        ))]),
+                        Err(_) => Err("Failed to update session metadata".to_string()),
+                    }
+                } else {
+                    Err("Failed to serialize TODO state".to_string())
+                }
+            }
+            Err(_) => Err("Failed to read session metadata".to_string()),
+        }
+    }
+
+    fn get_tools() -> Vec<Tool> {
+        fn create_schema(json_value: Value) -> Arc<Map<String, Value>> {
+            Arc::new(json_value.as_object().unwrap().clone())
+        }
+
+        vec![
+            Tool {
+                name: "todo_read".into(),
+                description: Some("Read the current TODO list content".into()),
```

Comment:
description needs to be expanded. Same for the write tool below.

Reply:
this is a sketch - the idea is to look at the architecture